### PR TITLE
Add SAM D21 family

### PR DESCRIPTION
**Pull request by 0xdec**
Add Microchip SAM D21/DA1 family (everything except the chip-scale packages).

[Datasheet](http://ww1.microchip.com/downloads/en/DeviceDoc/SAM_D21_DA1_Family_Data%20Sheet_DS40001882E.pdf)

- ATSAMD21E15A-A ATSAMD21E15A-A
- ATSAMD21E15A-M ATSAMD21E15A-M
- ATSAMD21E15L-A ATSAMD21E15L-A
- ATSAMD21E15L-M ATSAMD21E15L-M
- ATSAMD21G15A-A ATSAMD21G15A-A
- ATSAMD21G15A-M ATSAMD21G15A-M
- ATSAMD21G16L-M ATSAMD21G16L-M
- ATSAMD21J15A-A ATSAMD21J15A-A
- ATSAMD21J15A-M ATSAMD21J15A-M
- ATSAMD21J15B-C ATSAMD21J15B-C

---

All contributions to the kicad library must follow the [KiCad library convention](http://kicad-pcb.org/libraries/klc/)

Thanks for creating a pull request to contribute to the KiCad libraries! To speed up integration of your PR, please check the following items:

- [x] Provide a URL to a datasheet for the symbol(s) you are contributing
- [x] An example screenshot image is very helpful
- [x] Ensure that the associated footprints match the [official footprint library](https://github.com/kicad/kicad-footprints)
  - A new fitting footprint must be submitted if the library does not yet contain one.
- [x] If there are matching footprint PRs, provide link(s) as appropriate
- [x] Check the output of the Travis automated check scripts - fix any errors as required
- [x] Give a reason behind any intentional library convention rule violation.

### CODE REVIEW

**Comment by cpresser** [comment]
This seems to be a duplicate of https://github.com/KiCad/kicad-symbols/pull/2336
I have not taken a closer look at either PR yet, so right now I have no clue if one is 'better' than another.
We hat a addtion of other SAMD parts recently (https://github.com/KiCad/kicad-symbols/pull/2118). Some of the stylistic questions have already been discussed there. I think we should them also apply here, minus the bikeshedding :)

**Comment by 0xdec** [comment]
@cpresser this is #2336... I see some PRs for the SAMC family but not SAMD.

I read over #2118 pretty thoroughly while creating this PR.

**Comment by cpresser** [comment]
I will most likely review and hopefully merge this soon. 
Can you please do incremental commits starting from now on? That helps me to track changes/things that I already reviewed.

**Comment by 0xdec** [comment]
Sure thing! In past repos people have preferred force-pushing so I got in the habit of that. I just realized my pin lengths are 200mil when they should be 100mil, I'll fix that in the next couple hours.

**Comment by 0xdec** [comment]
@cpresser fixed the pin lengths

**Comment by 0xdec** [comment]
@cpresser any updates?

**Comment by cpresser** [comment]
Sorry for the huge delay. I am quite busy with moving our Hackerspace to a new location. This takes a insane amount of time out of my budget.
Anyway, I will make some time for the pending PRs on Github soon.

File: MCU_Microchip_SAMD.dcm [hunk 1]
```diff
@@ -78,12 +78,330 @@ K 32-bit ARM Cortex-M0+ MCU Microcontroller
...
+$ENDCMP
+#
+$CMP ATSAMD21E15A-M
+D SAM D21 Microchip SMART ARM-based Flash MCU, 48Mhz, 32K Flash, 32K SRAM, QFN-32
+K 32-bit ARM Cortex-M0+ MCU Microcontroller
+F http://ww1.microchip.com/downloads/en/DeviceDoc/SAMD21-Family-DataSheet-DS40001882D.pdf
+$ENDCMP
+#
+$CMP ATSAMD21E15B-A
+D SAM D21 Microchip SMART ARM-based Flash MCU, 48Mhz, 32K Flash, 32K SRAM, TQFP-32
```
Comment (cpresser) [inline MCU_Microchip_SAMD.dcm]: This `B` variant has the same description as the `A` variant.

**Comment by cpresser** [comment]
I will update this post with my review results. So please check if there are any edits.

I am not sure if we should include aliases for the device variants `B`.
Some of the variants have functional changes, others not. So I am not quite sure how to apply `S2.2`. Anyway, the description of different parts should differ. Currently they are the same.

Those are the new symbols:
* TQFP-32
  * `ATSAMD21E15A-A`
    * [x] Pin names and numbers are okay
    * [x] Footprint and filters
    * [ ] B-Variant description is same as A-Variant
  * `ATSAMD21E15L-A`
* QFN-32
  * `ATSAMD21E15A-M`
    * [ ] TODO
  * `ATSAMD21E15L-M`
    * [ ] TODO
* TQFP-48
  * `ATSAMD21G15A-A`
    * [ ] TODO
* QFN-48
  * `ATSAMD21G15A-M`
    * [ ] TODO
  * `ATSAMD21G16L-M`
    * [ ] TODO
* TQFP-64
  * `ATSAMD21J15A-A`
    * [ ] TODO
* QFN-64
  * `ATSAMD21J15A-M`
    * [ ] TODO
* UFBGA-64
  * `ATSAMD21J15B-C`
    * [ ] TODO

**Comment by cpresser** [comment]
@0xdec The review is done for now. I will do another round once you do some fixes. Overall this is a really good contribution. There are only minor issues. Keep up the good work!

**Comment by 0xdec** [comment]
> I am not sure if we should include aliases for the device variants `B`.
> Some of the variants have functional changes, others not. So I am not quite sure how to apply `S2.2`. Anyway, the description of different parts should differ. Currently they are the same. This applies to multiple parts/aliases. I will not add this as a checkbox item for each part.

I feel as though the variants have fairly specific changes that would require reading the datasheet, but don't really change the core functionality of the device. The symbol is unchanged for most of the variants (L excluded), so it seems to me that there's no reason to make an entirely new identical symbol.

>       * `ATSAMD21E15A-A`
>         
>         * [ ]  B-Variant and D-Variant description is same as A-Variant

The descriptions for each variant are rather verbose:

* A = Default Variant
* B = Added RWW support for 32 KB and 64 KB memory options
* C = Silicon revision F for WLCSP45 package option
* L  = Pinout optimized for Analog and PWM
* D = Silicon Revision G with RWW Support in 128KB memory options

Do you think it's worth it to try to simplify this description for the variant description? Or just include the variant name in the description?

>       * `ATSAMD21E15L-A`
>         
>         * [ ]  `VDDIO` pin not on same position as in `A`-Variant, thus not a drop-in-replacement
> 
>       * `ATSAMD21E15L-M`
>         
>         * [ ]  `VDDIO` pin not on same position as in `A`-Variant, thus not a drop-in-replacement

This is because the `A` variant does not have pin `VDDIO`, only `VDDIN`. These pins have different functions, but their position is consistent if you look at e.g. `ATSAMD21G15A-A`, which they are drop-in compatible with.

>       * `ATSAMD21E15A-M`
>         
>         * [x]  Filter should contain EP size

Done.

See above.

>       * `ATSAMD21G15A-A`
>         
>         * [x]  `VDDIO` pins should be unstacked since each one needs dedicated decoupling
> 
>       * `ATSAMD21G15A-M`
>         
>         * [x]  VDDIO-Pin-Stack as above
> 
>       * `ATSAMD21G16L-M`
>         
>         * [x]  VDDIO-Pin-Stack as above

Done, although it looks kinda ugly. Maybe the symbol body should be widened...

**Comment by 0xdec** [comment]
@cpresser oops looks like I made those changes while you updated your review...will look at the new stuff now. And thank you, I will!

**Comment by 0xdec** [comment]
Fixed almost everything, except descriptions of different variants. Let me know what you think about that.

**Comment by 0xdec** [comment]
@cpresser I think I addressed all the issues

**Comment by 0xdec** [comment]
As for the discussion about variant B, it seems that variants B, C, and D are simply silicon revisions that add RWW support to sections of the flash memory. I would argue that this is the same level of functional difference as different SRAM or FLASH sizes, so it shouldn't warrant a new symbol. The L variant is a bit different, it seems to contain the same silicon changes that add RWW support, as well as change which peripherals are available, and the device pinout.

**Comment by 0xdec** [comment]
Added the SAM DA1 family as well, since they're almost identical

**Comment by cpresser** [comment]
TIL: approving one commit out of 11 marks the whole thing as approved.

Thanks for the changes, they are good:
- [x] Descriptions (SRAM, Versions)
- [x] VDDIO vs. VDDIN thanks for the clarification. 
- [x] Unstacking the Pins
- [ ] Why are the DA1 part not aliases as well? They share the same pin-out as the D21!?

To be honest, I would rather have the DA1 parts in a dedicated PR. Now this one is insanely big. Which makes it pretty hard to properly review. Also the force-push is something that makes the review harder - now I have to check which ones of the individual commits was forced.
Please just add commits on top of the first one for future PRs.

**Comment by 0xdec** [comment]
I figured DA1 might be a separate PR, even though they're identical. Do you think it would be okay to just have them as aliases? I literally just copied them and changed the descriptions...

As for force-push, sorry about that. In past repos I've been requested to do that whenever I rebase, to avoid a merge commit. All my force-pushes have been only for rebasing on master, but I see your point. Will no longer force push from now on.

**Comment by cpresser** [comment]
> I figured DA1 might be a separate PR, even though they're identical. Do you think it would be okay to just have them as aliases? I literally just copied them and changed the descriptions...

I don't see a reason to not have them as alias. Am I missing something?

Would you undo the last commit in this branch? Then I will merge this one and we can work on the DA1 parts in a dedicated PR. That PR will most likely be trivial since its only aliases.

> 
> As for force-push, sorry about that. In past repos I've been requested to do that whenever I rebase, to avoid a merge commit. All my force-pushes have been only for rebasing on master, but I see your point. Will no longer force push from now on.

Actually I prefer re-basing over merging since merging produces ugly commits.
I was not aware that there was a conflict that required re-basing, I must have missed that - sorry.
So go ahead and rebase if required. The important part is to add a comment that explains why the history was changed.

**Comment by 0xdec** [comment]
Alias seems fine to me! I'll make a new PR for SAM DA1. Went ahead and un-did the last commit.

And sounds good on rebasing, will remember for future.

**Comment by cpresser** [comment]
Thanks for the good work! :+1:

**Comment by 0xdec** [comment]
@cpresser finally got around to opening a new PR for SAM DA1: #2687

**Pull request merged**